### ISSUE TITLE
Replay stage does not pick a fork based on voting

### ISSUE THREAD

**Issue by aeyakovenko** [solana-labs/solana#3127]
#### Problem

Replay stage does not pick a fork based on voting.

#### Proposed Solution

Implement locktower selection

It looks straightforward to implement locktower in the current replay stage

* Bank forks is the current active set of forks.  Regardless of vote or frozen state, these forks are all cousins. 
```
1. add a bit  to indicate if a fork has been 
    voted or not (bank forks can hold this,
    or the progress map)
2. Find all the active forks that pass the threshold rule
3. Filter out forks that are locked out
4. for all the valid non-voted active forks
  b. for each stakers vote program state in the bank
     1. simulate the stakers vote to a local copy of the vote program state (don't commit it to the bank!).
     2. add up the resulting total lockout in the entire bank.
  c. pick the fork with the highest total lockout.
  d. secondary rule: latest one by slot height.  
  e. TODO: Count network fee and use that as the secondary rule

```
tag: @carllin @rob-solana @garious

**Comment by garious** [comment]
Does Beacons need to be blocked on this? Why not pick the bank with the highest slot height and switch to Locktower in Grandview?

**Comment by aeyakovenko** [comment]
@garious stretch goal?

**Comment by garious** [comment]
Sure. Only blockers in Beacons now.

**Comment by carllin** [comment]
@aeyakovenko. Some questions:

1) Slightly confused about 4. b. 1., what are we simulating? I thought all the votes we needed to consider were already in the VoteState by that point. 

2) How is lockout calculated on a fork? Is it just the sum of the lockouts for each staker at the latest bank for that fork? Do we need to weight by length of the chain in any way?

**Comment by aeyakovenko** [comment]
@carllin
1. The Vote state in each fork can only reflect votes for the ancestors, and not the fork itself.  So we need to apply the vote for that fork and check if that fork pops more votes than it generates.
2. lockouts * stake weight

**Comment by carllin** [comment]
Also, is the definition of the "threshold rule" based on total lockout or remaining lockout depth. For instance, if I observe at slot 100 that a validator has a vote stack of depth 6 (lockout of 64), but that vote stack is going to expire at slot 101, do I still count that validator as having a lockout of 64?

**Comment by aeyakovenko** [comment]
@carllin, doesn’t matter so much, but it’s probably more accurate to use the simulated set. 

The reason it doesn’t matter is that the rule will fail on the next block and that fork will get unrolled.

**Comment by garious** [comment]
@aeyakovenko, @carllin,  what's the latest code do? Should we close this ticket?

**Comment by pgarg66** [comment]
Locktower is integrated in v0.12
